Harris Economic Order Quantity (EOQ) 

`y` is the optimal order quantity

`m` is the units per month

`s` is the setup cost of an order

`c` is the unit price of items in the stock

In [14]:
from dataclasses import dataclass
import numpy as np

@dataclass
class Inputs:
    """Represents a vector of model inputs.
    
    :attribute m: units used per month, or "movement" of stock
    :attribute s: setup cost of an order
    :attribute c: unit price of items in the stock
    """
    m: float 
    c: float
    s: float

def optimal_order_quantity(x: Inputs) -> int:
    """
    Economic order quantity function defined by Harris, 1990. 
    
    The number 240 is an "interest charge" per unit, per month, 
    corresponding to a yearly interest rate of 10 percent.
    
    :param x: dataclass representing a vector of model inputs. 
    :return: optimal order quantity
    """
    return np.floor(np.sqrt(240 * x.m * x.s / x.c)).astype(int)

y = optimal_order_quantity(Inputs(m=100.0, s=5.0, c=1.0))
print(y)

346


## Sensitivity Analysis with Finite Differences
### One-at-a-Time (OAT) Methods

We assign a base case `x0` and a sensitivity case `x_plus` to the model 
inputs. 

We change each model input, one at a time, by 10%. Then, we compare the 
resulting delta_y outputs (one for each changed parameter). 

In [21]:
x0 = Inputs(m=1230, c=0.0135, s=2.15)
y0 = optimal_order_quantity(x0)
x_plus_m = Inputs(m=1353, c=0.0135, s=2.15)
x_plus_s = Inputs(m=1230, c=0.01485, s=2.15)
x_plus_c = Inputs(m=1230, c=0.0135, s=2.365)
x_plus = Inputs(m=1353, c=0.01485, s=2.365)

delta_y_m = optimal_order_quantity(x_plus_m) - y0
delta_y_s = optimal_order_quantity(x_plus_s) - y0
delta_y_c = optimal_order_quantity(x_plus_c) - y0

deltas = [delta_y_m, delta_y_s, delta_y_c]
print("Base case:", y0)
print("Changes in model output with respect to m, s, and c:\n", deltas)

Base case: 6856
Changes in model output with respect to m, s, and c:
 [335, -319, 335]


This indicates that when `m` is increased by 10%, the model output increases
 by 335 units. When `s` is increased, the model output decreases by 319 
 units. 

In [24]:
delta_y_plus = optimal_order_quantity(x_plus) - y0
sum_of_delta = np.sum(deltas)

print("Change in y from x0 to x_plus:", delta_y_plus)
print("Sum of individual changes:", sum_of_delta)

Change in y from x0 to x_plus: 335
Sum of individual changes: 351


**Note that these are not equal**. This is due to the fact that this method of 
sensitivity analysis does not take interaction effects into account. 

## Scenario Decomposition
We start with a base case `x0`, best case `x_plus`, and worst case `x_minus`. 